lib/testing/matrix.js

```diff
@@ -0,0 +1,185 @@
+    *                      ex: for a hashmap with 'auth' as key and 'v4' as
+    *                      value, if we found a matrix with this following
+    *                      paramater, it will call the special callback
+    * @argument {GenerateMatrixCallback} callbackException - Callback who the
+    *                       execute method will call when
+    *                       this exception is found.
+    *
+    * @returns {TestMatrix} Return current instance of Matrix object
+    */
+    if(exception, callbackException) {
+        const pair = {
+            key: exception,
+            callback: callbackException,
+        };
+        this.listOfExceptions.push(pair);
+
+        return this;
+    }
+
+    /**
+    * Return the parameters of the matrix into a Json format
+    *
+    * @returns {string} JSON format of parameters
+    */
+    serialize() {
+        return this.paramsToString;
+    }
+
+    /**
+    * Generate the matrix and call all callback enter on @generate and
+    * if methods. Must be run as the father matrix for running all the test
+    *
+    * @returns {undefined}
+    */
+    execute() {
+        /**
+        * Check if @generate method has been already called
+        */
+        if (this.callback) {
+            /**
+            * Generate matrix element by element, we call the @execute
+            * method until that we specialize all element send
+            * by generate method
+            */
+            if (this.elementsToSpecialize.length !== 0) {
+                /**
+                * Make copy for restore object in the end of recursive function
+                */
+                const elementToSpecialize = this.elementsToSpecialize[0];
+                const copyElement = this.params[elementToSpecialize];
+
+                this.elementsToSpecialize.shift();
+                if (!(elementToSpecialize in this.params)) {
+                    const errMessage = 'No key of specialization found: ';
+                    throw new Error(errMessage + elementToSpecialize);
+                } else {
+                    copyElement.forEach(currentElement => {
+                        this.params[elementToSpecialize] = currentElement;
+                        this.execute();
+                    });
+                }
+                this.params[elementToSpecialize] = copyElement;
+                this.elementsToSpecialize.push(elementToSpecialize);
+            } else {
+                /**
+                * Create copy for not modify the current instance of the object
+                * during the call of the callback function
+                */
+                const matrixGenerated = new TestMatrix(this.params);
+                /**
+                * Search if the matrix we generated is an exception
+                */
+
+                const callFunction = (matrixFather, matrixChild, callback) => {
+                    const result = Object.keys(matrixChild.params)
```

Comment:
Doing this means that we need to ensure that the parameter's potential values do not contain objects, when creating the Matrix, right ?